python/lsst/pipe/tasks/diffractionSpikeMask.py

```diff
@@ -0,0 +1,436 @@
+
+        Parameters
+        ----------
+        x, y : `float`
+            Coordinates of the source to be masked.
+        r : `float`
+            Expected length of a diffraction spike for a source with a
+            magnitude
+
+        Returns
+        -------
+        mask : `lsst.afw.image.Mask`
+            A mask plane centered on the single source being modeled.
+        """
+        polygons = []
+        for angle in self.angles:
+            # Model the diffraction spike as an isosceles triangle with a long
+            # side along the spike and a short base. For efficiency, model the
+            # diffraction spike in the opposite direction at the same time,
+            # so the overall shape is a narrow diamond with equal length sides.
+            xLong = math.cos(np.deg2rad(angle))*r
+            yLong = math.sin(np.deg2rad(angle))*r
+            xShort = -math.sin(np.deg2rad(angle))*r/self.config.spikeAspectRatio
+            yShort = math.cos(np.deg2rad(angle))*r/self.config.spikeAspectRatio
+
+            corners = [lsst.geom.Point2D(x + xLong, y + yLong),
+                       lsst.geom.Point2D(x + xShort, y + yShort),
+                       lsst.geom.Point2D(x - xLong, y - yLong),
+                       lsst.geom.Point2D(x - xShort, y - yShort)]
+            polygons.append(afwGeom.Polygon(corners))
+        # Combine all of the polygons into a single region
+        polygon = polygons[0]
+        for poly in polygons[1:]:
+            polygon = polygon.unionSingle(poly)
+        bbox = lsst.geom.Box2I(polygon.getBBox())
+        polyImage = polygon.createImage(bbox).array
+        mask = lsst.afw.image.Mask(bbox)
+        mask.array[polyImage > 0] = mask.getPlaneBitMask(self.config.spikeMask)
+        return mask
+
+    def set_diffraction_angle(self, exposure):
+        """Calculate the angle of diffration spikes on the image given the
+        camera rotation.
+
+        Parameters
+        ----------
+        exposure : `lsst.afw.image.Exposure`
+            The exposure to calculate the expected angle of diffraction spikes.
+        """
+        angle = (exposure.visitInfo.boresightParAngle.asDegrees()
+                 - exposure.visitInfo.boresightRotAngle.asDegrees()
+                 + self.config.diffractAngle)
+        self.angles = np.array([angle, angle + 90]) % 360
+
+    def calculateReferenceRadius(self, magnitudes):
+        """Calculate the size of the region to mask for each bright star.
+
+        Parameters
+        ----------
+        magnitudes : `numpy.ndarray`
+            Magnitudes of the bright stars.
+
+        Returns
+        -------
+        radii : `numpy.ndarray`
+            Array of radius values for the given magnitudes.
+        """
+        # In pixels
+        radii = [10**(self.config.magSlope*magnitude + self.config.magOffset) for magnitude in magnitudes]
+        return np.array(radii)
+
+    def extractMagnitudes(self, refCat, filterLabel):
```

Comment:
I hate that this code needs to be duplicated from elsewhere, but I'm not sure what else to do.